Product Gallery block: Migrate to new Interactivity API store

thealexandrelara opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

Fixes #11656 

## What

Migrates the Product Gallery block and its inner blocks (Pager, Thumbnails, Large Image blocks) to the new store() API.

Resources:

- https://github.com/woocommerce/woocommerce-blocks/issues/11065
- https://github.com/woocommerce/woocommerce-blocks/pull/11071
- https://github.com/woocommerce/woocommerce-blocks/pull/11558
- https://github.com/WordPress/gutenberg/discussions/53586

## Why

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->

Because we are switching from the old `store()` API to the new `store()` API, and we need to migrate all the blocks at once.

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Three," "Twenty-twenty Four," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. Inside the Page editor, click on the '+' button to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block. Click on it to add the block to the template.
7. Click on Save;
8. Visit a product page and make sure the block and its inner blocks are working as expected.

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

-------

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->

thealexandrelara committed:
Migrate Product Gallery block to new Interactivity API store

luisherranz committed:
Merge branch 'interactivity-api-refactor-product-button-with-new-store-api' into refactor/11656-migrate-product-gallery-block-to-new-interactivity-store-api

luisherranz committed:
Fix some references

luisherranz committed:
Add missing data-wc-interactive

luisherranz committed:
Fix an additional namespace

luisherranz committed:
Remove unnecessary click handler

luisherranz committed:
Dialog working

luisherranz committed:
Refactor action names

luisherranz committed:
Reindex PHP array

There was some missing indexes, which turned the array into an object in JS.

luisherranz committed:
Remove unused event handlers

luisherranz committed:
Move next/previous logic to external function

luisherranz committed:
Move StorePart util to the types folder

luisherranz committed:
Rename namespace to `woocommerce/product-gallery`

luisherranz committed:
Undo product collection namespace renaming

luisherranz committed:
Remove unnecessary namespace

luisherranz committed:
Don't hide the large image on page load

luisherranz committed:
Minor refactorings

luisherranz commented:
Hey @thealexandrelara, I've merged the changes in `trunk`, fixed the bug we discovered yesterday, and refactored the rest of the logic.

I took a video to explain the whole refactoring, including what we did yesterday as well.


<div>
    <a href="https://www.loom.com/share/76a974c4901646c295ddc0187d44180a">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/76a974c4901646c295ddc0187d44180a-with-play.gif">
    </a>
  </div>

https://www.loom.com/share/76a974c4901646c295ddc0187d44180a

Please review everything again because now there are many more changes in the code. Also, take a look at the variations because I didn't check them, and at the points in the opened dialog which are wrong.

I'll take a look at how to work with global event listeners because I don't like much that you can't easily pass down a scope, but we can merge this as it is.

thealexandrelara committed:
Fix eslint error

thealexandrelara committed:
Fix php cs errors with spacing and double arrows alignment

thealexandrelara commented:
> Hey @thealexandrelara, I've merged the changes in `trunk`, fixed the bug we discovered yesterday, and refactored the rest of the logic.
> 
> I took a video to explain the whole refactoring, including what we did yesterday as well.
> 
> [ ![](https://camo.githubusercontent.com/de57fba98a54e3dffcf68c1586badfcdd805b2ee788c103b871832e4d1c124af/68747470733a2f2f63646e2e6c6f6f6d2e636f6d2f73657373696f6e732f7468756d626e61696c732f37366139373463343930313634366332393564646330313837643434313830612d776974682d706c61792e676966) ](https://www.loom.com/share/76a974c4901646c295ddc0187d44180a)
> https://www.loom.com/share/76a974c4901646c295ddc0187d44180a
> 
> Please review everything again because now there are many more changes in the code. Also, take a look at the variations because I didn't check them, and at the points in the opened dialog which are wrong.
> 
> I'll take a look at how to work with global event listeners because I don't like much that you can't easily pass down a scope, but we can merge this as it is.

Hey @luisherranz, thanks a lot for resolving the conflicts and providing this walkthrough of the changes. I've just reviewed and tested the block, and everything is working as expected. 

Regarding the issue with opening the dialog and the Pager block, this is something that is being addressed at https://github.com/woocommerce/woocommerce-blocks/pull/11716. I've also merged a fix for the PHP Code Standards errors.

luisherranz commented:
Nice 🙂 

Feel free to mark this as ready for review once you've fixed everything, @thealexandrelara.

thealexandrelara committed:
Disable no-use-before-define rule for eslint

thealexandrelara committed:
Disable @typescript-eslint/ban-types rule for eslint

thealexandrelara committed:
Merge branch 'interactivity-api-refactor-product-button-with-new-store-api' into refactor/11656-migrate-product-gallery-block-to-new-interactivity-store-api

thealexandrelara committed:
Fix parsed context error in e2e tests

thealexandrelara committed:
Fix context parser for Thumbnail image

thealexandrelara commented on .eslintrc.js:
```diff
@@ -241,7 +241,7 @@ module.exports = {
 				'@typescript-eslint/no-explicit-any': 'error',
 				'no-use-before-define': 'off',
-				'@typescript-eslint/no-use-before-define': [ 'error' ],
+				'@typescript-eslint/no-use-before-define': 'off',
```
This is needed because in the new Interactivity Store API we need to access the `state`,  `actions` and other components from the store before the store is defined, as you can see here: https://github.com/woocommerce/woocommerce-blocks/pull/11721/files#diff-40099b9b90dd057ec2594e0ceee370c105bbf64c58b94396e278b2a7819b0abdR155

Basically, we destructure the return of the `store` call and use the `state` before that declaration in here:
https://github.com/woocommerce/woocommerce-blocks/pull/11721/files#diff-40099b9b90dd057ec2594e0ceee370c105bbf64c58b94396e278b2a7819b0abdR51

As you can see from the above examples, the state is being used on Line 51, but it is only defined in line 155.

thealexandrelara replied:
Solved by moving the store to the top and https://github.com/woocommerce/woocommerce-blocks/pull/11721#issuecomment-1808961451.

luisherranz commented:
You can move the `const { state } = store('...')` call to the top if you prefer: https://github.com/woocommerce/woocommerce-blocks/commit/7a1dd1ad11d2d265c9dbd184b230f90fec1e4fc0

You can call `store()` as many times as you want, and it will always return the same object for the same namespace! 🙂

thealexandrelara committed:
Move store to the top of the frontend file

thealexandrelara commented:
> You can move the `const { state } = store('...')` call to the top if you prefer: [7a1dd1a](https://github.com/woocommerce/woocommerce-blocks/commit/7a1dd1ad11d2d265c9dbd184b230f90fec1e4fc0)
> 
> You can call `store()` as many times as you want, and it will always return the same object for the same namespace! 🙂

Ohh nice! Thank you for your tip, I just pushed the changes with your suggestion!

thealexandrelara committed:
Add interactivity api utils to the @woocommerce/utils alias

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into refactor/11656-migrate-product-gallery-block-to-new-interactivity-store-api

DAreRodz commented:
I've updated the branch base to [interactivity-api-new-store-api](https://github.com/woocommerce/woocommerce-blocks/tree/interactivity-api-new-store-api) and merged it to update and resolve conflicts.

EDIT: changed the base branch back to [interactivity-api-refactor-product-button-with-new-store-api](https://github.com/woocommerce/woocommerce-blocks/tree/interactivity-api-refactor-product-button-with-new-store-api)

DAreRodz commented on src/BlockTypes/ProductGalleryLargeImage.php:
```diff
@@ -117,10 +117,10 @@ function( $carry, $key ) use ( $directives ) {
+			'class'                => 'wc-block-woocommerce-product-gallery-large-image__image',
+			'data-wc-bind--hidden' => '!state.isSelected',
+			'data-wc-watch'        => 'callbacks.scrollInto',
+			'data-wc-class--wc-block-woocommerce-product-gallery-large-image__image--active-image-slide' => 'state.isSelected',
```
This line was added during https://github.com/woocommerce/woocommerce-blocks/pull/11721/commits/4d6a0b0cd2390c63a2d9bcd79b8e32b8f6968f9c merge, although I'm not sure if it should be kept or not. 🤔

cc: @thealexandrelara

thealexandrelara replied:
Thank you for the ping, yes, it should be there. We are using it to control the currently active image.

DAreRodz committed:
Merge branch 'interactivity-api-refactor-product-button-with-new-store-api' into refactor/11656-migrate-product-gallery-block-to-new-interactivity-store-api

DAreRodz commented on assets/js/blocks/product-gallery/frontend.tsx:
```diff
@@ -1,221 +1,161 @@
+	productId: string;
 }
 
 interface Event {
 	keyCode: number;
 }
```
This wouldn't be necessary if the `KeyboardEvent` class is used as the type in the callback below. 👇

In addition, the `keyCode` property is deprecated―it should be replaced with `key` or `code`. See [KeyboardEvent/keyCode](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode).

DAreRodz commented on assets/js/blocks/product-gallery/frontend.tsx:
```diff
@@ -1,221 +1,161 @@
-			keyboardAccess: ( store: Store ) => {
-				const { context, actions } = store;
-				let allowNavigation = true;
+			const handleKeyEvents = ( event: Event ) => {
```
Instead of the custom interface, it would be better to use `KeyboardEvent` here, right? 🤔

DAreRodz commented on assets/js/blocks/product-gallery/frontend.tsx:
```diff
@@ -1,221 +1,161 @@
-						allowNavigation = true;
-					} );
+				// Check if the esc key is pressed.
+				if ( event.keyCode === Keys.ESC ) {
```
As mentioned above, `keyCode` is deprecated. It'd be worth considering replacing this (and other `event.keyCode` calls).

roykho replied:
@DAreRodz - thanks for the suggestions. I will do a follow up PR after merge.

roykho replied:
Wasn't too much work so I just pushed up the changes.

DAreRodz reviewed (commented):
I just left some suggestions that I think should be addressed (maybe in a subsequent PR 🙂).

roykho committed:
Replace deprecated event attribute

roykho reviewed (approved):
Everything tested nicely and the code looks pretty straight forward! Good job everyone!
